src/kvstore/comm.h

```diff
@@ -635,6 +656,302 @@ class CommDevice : public Comm {
   bool inited_;
 };
 
+#if MXNET_USE_NCCL
+class CommNCCL : public Comm {
+ public:
+  CommNCCL() {
+    inited_ = false;
+    pinned_ctx_ = Context::CPUPinned(0);
+  }
+
+  virtual ~CommNCCL() {
+    for (auto e : nccl_data_) {
+      cudaStreamDestroy(e.second.stream);
+      ncclCommDestroy(e.second.comm);
+    }
+  }
+
+  void Init(int key, const NDArrayStorageType stype, const TShape& shape,
+            int dtype = mshadow::kFloat32, Context pinned_ctx = Context::CPUPinned(0)) override {
+    if (stype == kDefaultStorage) {
+      sorted_key_attrs_.push_back(std::make_tuple(key, shape, dtype));
+    } else {
+      LOG(FATAL) << "NCCL KVStore does not support sparse storage type";
+    }
+  }
+
+  const NDArray& Reduce(int key, const std::vector<NDArray>& src,
+                        int priority) override {
+    // avoid extra copy for single device, but it may bring problems for
+    // abnormal usage of kvstore
+    if (src.size() == 1) {
+      return src[0];
+    }
+
+    if (!inited_) {
+      std::vector<Context> devs;
+      for (const auto& a : src) {
+        devs.push_back(a.ctx());
+      }
+      InitNCCL(devs);
+      InitMergeBuffer(devs);
+    }
+
+    std::vector<int> dev_ids;
+    for (auto e : src) {
+      dev_ids.push_back(e.ctx().dev_id);
+    }
+    std::sort(dev_ids.begin(), dev_ids.end());
+    CHECK(device_ids_ == dev_ids) << "NCCL KVStore supports only single set of devices";
```

Comment:
Yes. Handling multiple sets of devices can be done, but not with the structure imposed by the `Comm` class. Basically in order to keep the benefits of batching I need to ensure that the root for the reduction will be the same for the whole batch, but I know who participates only during the actual push/pull, not during Init, and all of the data structures are initialized only once during the first push. This BTW should also be checked in the `device` kvstore (and currently is not), otherwise you can do something like this:
```
>>> import mxnet as mx
>>> kv = mx.kv.create("device")
>>> shape = (2,3)
>>> kv.init(4, mx.nd.ones(shape))
>>> gpus = [mx.gpu(i) for i in range(2)]
>>> b = [mx.nd.ones(shape, gpu) for gpu in gpus]
>>> kv.push(4, b)
>>> a = mx.nd.zeros(shape)
>>> kv.pull(4, out = a)
>>> a
[[ 2.  2.  2.]
 [ 2.  2.  2.]]
<NDArray 2x3 @cpu(0)>
>>> gpus = [mx.gpu(i) for i in range(4)]
>>> 
>>> b = [mx.nd.ones(shape, gpu) for gpu in gpus]
>>> kv.push(4, b)
Segmentation fault
```